### DISCUSSION TITLE
macOS Sonoma fails installing PETSc

### DISCUSSION THREAD

**Discussion by fberonvera** [firedrakeproject/firedrake#3392] [Firedrake support]
My installation gets interrupted when attempting to install PETSc.  Hope somebody can get me some hint?  Thank you in advance.  -Francisco

Running firedrake-install
Installing command line tools...
Checking brew doctor...
Installing required packages via homebrew...
Required package 'gcc' is already installed via Homebrew.
Required package 'autoconf' is already installed via Homebrew.
Required package 'make' is already installed via Homebrew.
Required package 'automake' is already installed via Homebrew.
Required package 'cmake' is already installed via Homebrew.
Required package 'libtool' is already installed via Homebrew.
Required package 'boost' is already installed via Homebrew.
Creating firedrake venv in '/Users/fberon/firedrake'.
BLAS config is {'BLAS': '/Users/fberon/firedrake/src/petsc/default', 'OPENBLAS': '/Users/fberon/firedrake/src/petsc/default', 'LDFLAGS': '-L/usr/local/opt/gcc/lib/gcc/11'}
Configuration saved to /Users/fberon/firedrake/.configuration.json
Cloning firedrake

Failed to clone firedrake using ssh, falling back to https.
Successfully cloned repository firedrake.
Checking out branch master
Successfully checked out branch master
Updating submodules.
Successfully updated submodules.
Creating firedrake-update script.
Cloning the dependencies of firedrake
Cloning ufl

Failed to clone ufl using ssh, falling back to https.
Successfully cloned repository ufl.
Checking out branch master

...

Cloning the dependencies of PyOP2
Cloning loopy

Failed to clone loopy using ssh, falling back to https.
Successfully cloned repository loopy.
Checking out branch main
Successfully checked out branch main
Updating submodules.
Successfully updated submodules.
Installing petsc/
Depending on your platform, PETSc may take an hour or more to build!
Traceback (most recent call last):
  File "/Users/fberon/firedrake-install", line 1865, in <module>
    install("petsc/")
  File "/Users/fberon/firedrake-install", line 1055, in install
    build_and_install_petsc()
  File "/Users/fberon/firedrake-install", line 1167, in build_and_install_petsc
    check_call([python, "./configure", "PETSC_DIR={}".format(petsc_dir), "PETSC_ARCH={}".format(petsc_arch)] + petsc_options)
  File "/Users/fberon/firedrake-install", line 668, in check_call
    log.debug(subprocess.check_output(arguments, stderr=subprocess.STDOUT, env=os.environ).decode())
  File "/usr/local/Cellar/python@3.9/3.9.12/Frameworks/Python.framework/Versions/3.9/lib/python3.9/subprocess.py", line 424, in check_output
    return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
  File "/usr/local/Cellar/python@3.9/3.9.12/Frameworks/Python.framework/Versions/3.9/lib/python3.9/subprocess.py", line 528, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['/Users/fberon/firedrake/bin/python', './configure', 'PETSC_DIR=/Users/fberon/firedrake/src/petsc', 'PETSC_ARCH=default', '--with-debugging=0', '--download-netcdf', '--download-metis', '--download-scalapack', '--with-x=0', '--download-suitesparse', '--download-pastix', '--download-hwloc', '--CFLAGS=-Wno-implicit-function-declaration', '--LDFLAGS=-Wl,-ld_classic', '--download-mumps', '--download-hwloc-configure-arguments=--disable-opencl', '--with-shared-libraries=1', '--download-openblas', '--download-bison', '--with-c2html=0', '--download-ptscotch', '--with-zlib', '--download-mpich-device=ch3:sock', '--download-pnetcdf', '--download-mpich', '--download-mpich-configure-arguments=--disable-opencl', '--download-hypre', '--with-fortran-bindings=0', '--download-chaco', '--download-superlu_dist', "--download-openblas-make-options='USE_THREAD=0 USE_LOCKING=1 USE_OPENMP=0'", '--download-hdf5']' returned non-zero exit status 1.

Install log saved in /Users/fberon/firedrake-install.log

**Comment by fberonvera** [comment]
Attached is the firedrake-install.log:

[firedrake-install.log](https://github.com/firedrakeproject/firedrake/files/14188761/firedrake-install.log)

In addition:

fberon@MacBook[~]%  which python3
/usr/local/bin/python3

**Reply by connorjward** [reply]
Please can you also share `firedrake/src/petsc/configure.log`?

From looking at your logs you are failing to build bison, which I have not encountered before. I notice that `brew doctor` is raising a number of concerning looking warnings. Can you make sure that your homebrew is fully updated (`brew upgrade`) and also resolve the issues raised by `brew doctor` before trying again.

**Comment by fberonvera** [comment]
[configure.log](https://github.com/firedrakeproject/firedrake/files/14194212/configure.log)

**Reply by connorjward** [reply]
This includes the lines
```
checking for gcc... gcc
checking whether the C compiler works... noconfigure: error: in `/Users/fberon/firedrake/src/petsc/default/externalpackages/bison-3.8.2':
configure: error: C compiler cannot create executables
```
As mentioned, I think there is something wrong with your homebrew or command line tools. Please make sure both are up-to-date and that `brew doctor` raises no warnings.

**Reply by fberonvera** [reply]
Can this:

Warning: gcc 11.2.0_3 is already installed, it's just not linked.
To link this version, run:
  brew link gcc
fberon@MacBook[~]%  brew link gcc
Linking /usr/local/Cellar/gcc/11.2.0_3...
Error: Could not symlink include/c++/11
/usr/local/include/c++ is not writable.

be the issue?

**Reply by connorjward** [reply]
Perhaps. If the `brew doctor` issues cannot be resolved then a fresh install of homebrew may be sensible.

**Reply by fberonvera** [reply]
Done, but not tried to python3 firedrake-install yet.  brew doctor lists a number of 'Warning: Unbrewed...'  Is this to worry about?  No unliked things anymore.  brew is up to date.

**Reply by fberonvera** [reply]
[firedrake-install.log](https://github.com/firedrakeproject/firedrake/files/14196871/firedrake-install.log)
[configure.log](https://github.com/firedrakeproject/firedrake/files/14196878/configure.log)

**Reply by connorjward** [reply]
`brew doctor` is still raising a huge number of warnings. Please resolve these before running `firedrake-install` again.

**Reply by fberonvera** [reply]
Do you mean, for instance, I have to DELETE these files:

 690 Warning: Unbrewed '.la' files were found in /usr/local/lib.
 691 If you didn't put them there on purpose they could cause problems when
 692 building Homebrew formulae, and may need to be deleted.
 693
 694 Unexpected '.la' files:
 695   /usr/local/lib/libasan.la
 696   /usr/local/lib/libatomic.la
 697   /usr/local/lib/libgfortran.la
 698   /usr/local/lib/libgmp.la
 699   /usr/local/lib/libgomp.la
 700   /usr/local/lib/libitm.la
 701   /usr/local/lib/libmpc.la
 702   /usr/local/lib/libmpfr.la
 703   /usr/local/lib/libquadmath.la
 704   /usr/local/lib/libssp.la
 705   /usr/local/lib/libssp_nonshared.la
 706   /usr/local/lib/libstdc++.la
 707   /usr/local/lib/libsupc++.la

??

**Reply by connorjward** [reply]
If that is what homebrew suggests then yes.

**Comment by fberonvera** [comment]
Latest log files included.  I'm giving up...
[configure.log](https://github.com/firedrakeproject/firedrake/files/14202365/configure.log)
[firedrake-install.log](https://github.com/firedrakeproject/firedrake/files/14202370/firedrake-install.log)

**Reply by fberonvera** [reply]
Now (after deleting unbrewed files one by one; scary):
fberon@MacBook[~]%  brew doctor
Your system is ready to brew.

However:
[configure.log](https://github.com/firedrakeproject/firedrake/files/14209697/configure.log)

:-(

**Reply by connorjward** [reply]
The `configure.log` shows that PETSc built successfully. Could you share `firedrake-install.log` instead?

However, from looking at prior `firedrake-install.log` files I think your issue may be that you are not using the right Python when you run `firedrake-install`. You want to use a Python installed by Homebrew:

```
$ $(brew --prefix)/bin/python3 firedrake-install
```

But we seem to be getting somewhere!

**Reply by fberonvera** [reply]
But:
fberon@MacBook[~]%  which python3
/usr/bin/python3

I
rm firedrake-install.log

:-(

**Reply by fberonvera** [reply]
Oh, I see,
fberon@MacBook[~]%  brew --prefix
/usr/local

I'm sure I
brew install python@3.10

**Reply by connorjward** [reply]
Is it working? I think `brew install python` is sufficient. No need to specify a version number.

**Reply by fberonvera** [reply]
Latest logs, look better, but there's still are errors in there...

[configure.log](https://github.com/firedrakeproject/firedrake/files/14211507/configure.log)
[firedrake-install.log](https://github.com/firedrakeproject/firedrake/files/14211508/firedrake-install.log)

Compiled using:
fberon@MacBook[~]%  /usr/local/bin/python3 firedrake-install

Doctor says:
fberon@MacBook[~]%  brew doctor
Your system is ready to brew.

Sorry...

**Reply by connorjward** [reply]
You've hit issues with libsupermesh and rtree. @JDBetteridge was working on this recently. How old is the `firedrake-install` script that you downloaded? The error may go away with a freshly downloaded script.

**Comment by JDBetteridge** [comment]
You are hitting the issue that I documented here https://github.com/firedrakeproject/libsupermesh/blob/master/CMakeLists.txt#L161 I legitimately have no idea how to test for C++11 strings on a Mac. There is no canonical way to test anyway, my method works on Linux by querying the strings in the shared object (not ideal).

Can you try in your activated virtual environment:
```bash
cd $VIRTUAL_ENV/src/libsupermesh/build
$VIRTUAL_ENV/src/petsc/default/bin/cmake .. \
    -DBUILD_SHARED_LIBS=ON \
    -DMAKE_CXX_FLAGS="-D_GLIBCXX_USE_CXX11_ABI=0"
    -DCMAKE_INSTALL_PREFIX=$VIRTUAL_ENV \
    -DMPI_C_COMPILER=$VIRTUAL_ENV/bin/mpicc \
    -DMPI_CXX_COMPILER=$VIRTUAL_ENV/bin/mpicxx \
    -DMPI_Fortran_COMPILER=$VIRTUAL_ENV/bin/mpif90 \
    -DCMAKE_Fortran_COMPILER=$VIRTUAL_ENV/bin/mpif90 \
    -DMPIEXEC_EXECUTABLE=$VIRTUAL_ENV/bin/mpiexec
make
make install
firedrake-update
```

**Reply by fberonvera** [reply]
I'm using the one provided here:

https://www.firedrakeproject.org/download.html

?

**Reply by JDBetteridge** [reply]
You don't need to download anything new. Activate the virtual environment from you previous failed install and then run those commands.

**Comment by fberonvera** [comment]
This what I got up to the first error (ther was one before saying that cmake was not available in /Users/fberon/firedrake/src/petsc/default/bin/, so I sudo ln -s /usr/local/bin/cmake /Users/fberon/firedrake/src/petsc/default/bin/cmake):

(firedrake) fberon@MacBook[~/firedrake/src/libsupermesh/build]%  -DCMAKE_INSTALL_PREFIX=$VIRTUAL_ENV \
> -DMPI_C_COMPILER=$VIRTUAL_ENV/bin/mpicc \
> -DMPI_CXX_COMPILER=$VIRTUAL_ENV/bin/mpicxx \
> -DMPI_Fortran_COMPILER=$VIRTUAL_ENV/bin/mpif90 \
> -DCMAKE_Fortran_COMPILER=$VIRTUAL_ENV/bin/mpif90 \
> -DMPIEXEC_EXECUTABLE=$VIRTUAL_ENV/bin/mpiexec
-bash: -DCMAKE_INSTALL_PREFIX=/Users/fberon/firedrake: No such file or directory
(firedrake) fberon@MacBook[~/firedrake/src/libsupermesh/build]%  cd $VIRTUAL_ENV/src/libsupermesh/build
(firedrake) fberon@MacBook[~/firedrake/src/libsupermesh/build]%  $VIRTUAL_ENV/src/petsc/default/bin/cmake .. \
>     -DBUILD_SHARED_LIBS=ON \
>     -DMAKE_CXX_FLAGS="-D_GLIBCXX_USE_CXX11_ABI=0"
-- Disabling debugging
CMake Warning (dev) at CMakeLists.txt:31 (exec_program):
  Policy CMP0153 is not set: The exec_program command should not be called.
  Run "cmake --help-policy CMP0153" for policy details.  Use the cmake_policy
  command to set the policy and suppress this warning.

  Use execute_process() instead.
This warning is for project developers.  Use -Wno-dev to suppress it.

-- Git revision: 84becef14eb117defa49354116c04aa180471a07
--
-- MPI_Fortran_COMPILER=/Users/fberon/firedrake/src/petsc/default/bin/mpif90
-- CMAKE_Fortran_FLAGS= -ffree-line-length-none -std=f2008 -fPIC
-- CMAKE_Fortran_FLAGS_DEBUG=-g
-- CMAKE_Fortran_FLAGS_RELEASE=-O3 -DNDEBUG -O3
--
-- MPI_C_COMPILER=/Users/fberon/firedrake/src/petsc/default/bin/mpicc
-- CMAKE_C_FLAGS=-L/Users/fberon/firedrake/src/petsc/default/lib -I/Users/fberon/firedrake/src/petsc/default/include -std=c99 -fPIC
-- CMAKE_C_FLAGS_DEBUG=-g
-- CMAKE_C_FLAGS_RELEASE=-O3 -DNDEBUG
--
-- MPI_CXX_COMPILER=/Users/fberon/firedrake/src/petsc/default/bin/mpicxx
-- CMAKE_CXX_FLAGS= -std=c++11 -fPIC
-- CMAKE_CXX_FLAGS_DEBUG=-g
-- CMAKE_CXX_FLAGS_RELEASE=-O3 -DNDEBUG
--
-- MPIEXEC_EXECUTABLE=/Users/fberon/firedrake/src/petsc/default/bin/mpiexec
--
-- Python_VERSION_MAJOR 3
-- Python_VERSION_MINOR 11
-- Found libspatialindex shared library: /Users/fberon/firedrake/lib/python3.11/site-packages/rtree/lib/libspatialindex.dylib
-- Found libspatialindex headers: /Users/fberon/firedrake/lib/python3.11/site-packages/rtree/include
-- Configuring done (3.6s)
-- Generating done (0.1s)
-- Build files have been written to: /Users/fberon/firedrake/src/libsupermesh/build
(firedrake) fberon@MacBook[~/firedrake/src/libsupermesh/build]%      -DCMAKE_INSTALL_PREFIX=$VIRTUAL_ENV \
>     -DMPI_C_COMPILER=$VIRTUAL_ENV/bin/mpicc \
>     -DMPI_CXX_COMPILER=$VIRTUAL_ENV/bin/mpicxx \
>     -DMPI_Fortran_COMPILER=$VIRTUAL_ENV/bin/mpif90 \
>     -DCMAKE_Fortran_COMPILER=$VIRTUAL_ENV/bin/mpif90 \
>     -DMPIEXEC_EXECUTABLE=$VIRTUAL_ENV/bin/mpiexec
-bash: -DCMAKE_INSTALL_PREFIX=/Users/fberon/firedrake: No such file or directory

**Reply by connorjward** [reply]
I think a backslash is missing from the cmake instruction so it's getting run as two instructions rather than one.

**Comment by fberonvera** [comment]
Sorry again.  I added \ and when got make (which doesn't exist in /Users/fberon/firedrake/src/petsc/default/bin/ so I sudo ln -s /usr/local/bin/cmake /Users/fberon/firedrake/src/petsc/default/bin/cmake) it run until here:

[ 47%] Linking CXX executable test_element_intersector
Undefined symbols for architecture x86_64:
  "__ZN5Tools11PropertySet11setPropertyENSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEERKNS_7VariantE", referenced from:
      __ZN12libsupermesh5RTreeC1ERKiPKdS2_S2_PS1_ in libsupermesh.a(R-Tree_Intersection_Finder_C++.cpp.o)
ld: symbol(s) not found for architecture x86_64
collect2: error: ld returned 1 exit status
make[2]: *** [test_element_intersector] Error 1
make[1]: *** [CMakeFiles/test_element_intersector.dir/all] Error 2
make: *** [all] Error 2

**Comment by connorjward** [comment]
Let's try a different approach. Can you see if your "command line tools" are fully up-to-date. You can update them by:
```
Use "Software Update" in System Settings or the softwareupdate command line interface to install updates
```

Basically the problem that you are currently experiencing is that there is still something slightly wrong with the C++ compiler on your machine. It doesn't seem to support a new-ish type of string. Hopefully if you update command line tools the error will go away and you can run `$(brew --prefix)/bin/python3 firedrake-install` and it will "just work".

I am surprised that you are hitting this issue. I don't think any of our other MacOS users have run into this.

**Comment by fberonvera** [comment]
What headache.  I upgraded to Xcode 15.2 beta and the corresponding command line tools.  However:

fberon@MacBook[/]%  brew doctor
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: A newer Command Line Tools release is available.
Update them from Software Update in System Settings.

If that doesn't show you any updates, run:
  sudo rm -rf /Library/Developer/CommandLineTools
  sudo xcode-select --install

Alternatively, manually download them from:
  https://developer.apple.com/download/all/.
You should download the Command Line Tools for Xcode 15.1.

**Comment by fberonvera** [comment]
I  did the above and got this:
[configure.log](https://github.com/firedrakeproject/firedrake/files/14217682/configure.log)
[firedrake-install.log](https://github.com/firedrakeproject/firedrake/files/14217686/firedrake-install.log)
Not sure if this is Sonoma or what.  Fortunately in a desktop running Ventura I guess did get installed well.

**Reply by connorjward** [reply]
@JDBetteridge  I'm stumped. Do you have any further ideas?
